src/main/java/org/tugraz/sysds/runtime/data/LibTensorAgg.java

```diff
@@ -0,0 +1,396 @@
+		KahanPlus akplus = KahanPlus.getKahanPlusFnObject();
+
+		//incl implicit nnz maintenance
+		for(int i=0, ix=0; i<m; i++, ix+=n)
+			for(int j=0; j<n-1; j++) {
+				buffer._sum = aggVal.get(new int[]{i, j});
+				buffer._correction = aggVal.get(new int[]{i, n - 1});
+				akplus.execute(buffer, a[ix + j], a[ix + j + 1]);
+				aggVal.set(new int[]{i, j}, buffer._sum);
+				aggVal.set(new int[]{i, n - 1}, buffer._correction);
+			}
+		// TODO check sparsity
+	}
+
+	/**
+	 * Add two partial aggregations together.
+	 * @param aop the aggregation operator
+	 * @param out the tensor-block which contains partial result and should be increased to contain sum of both results
+	 * @param partout the tensor-block which contains partial result and should be added to other partial result
+	 */
+	private static void aggregateFinalResult(AggregateOperator aop, TensorBlock out, TensorBlock partout ) {
+		AggregateOperator laop = aop;
+
+		//TODO special handling for mean where the final aggregate operator (kahan plus)
+		// is not equals to the partial aggregate operator
+		if( aop.increOp.fn instanceof Mean ) {
+			laop = new AggregateOperator(0, KahanPlus.getKahanPlusFnObject(), aop.correctionExists,
+					aop.correctionLocation);
+		}
+
+		//incremental aggregation of final results
+		if( laop.correctionExists )
+			out.incrementalAggregate(laop, partout);
+		else
+			throw new NotImplementedException();
+			//out.binaryOperationsInPlace(laop.increOp, partout);
+	}
+
+	/**
+	 * SUM, opcode: uak+, dense input.
+	 *
+	 * @param a input dense block
+	 * @param c output dense block with correction
+	 * @param kbuff kahan buffer
+	 * @param kplus kahan plus object
+	 * @param rl row lower index
+	 * @param ru row upper index
+	 */
+	private static void d_uakp( DenseBlock a, DenseBlock c, KahanObject kbuff, KahanPlus kplus, int rl, int ru ) {
+		final int bil = a.index(rl);
+		final int biu = a.index(ru-1);
+		if (a instanceof DenseBlockBool || a instanceof DenseBlockLBool) {
+			kbuff._sum = a.countNonZeros(rl, ru, 0, a.getCumODims(0));
+		} else if (a instanceof DenseBlockFP64 || a instanceof DenseBlockLFP64) {
+			for (int bix = bil; bix <= biu; bix++) {
+				double[] values = a.valuesAt(bix);
+				int lpos = (bix==bil) ? a.pos(rl) : 0;
+				int len = (bix==biu) ? a.pos(ru - 1) - lpos + a.getCumODims(0) : a.blockSize(bix) * a.getCumODims(0);
+				sum(values, lpos, len, kbuff, kplus);
+			}
+		} else {
+			// TODO special case String: currently converted to double, should be concatenated?
+			// Less readable and makes use of how row-column-getter actually works, but is a lot faster than
+			// using the index[]-getter.
+			long sum = 0;
+			for (int bix = bil; bix <= biu; bix++) {
+				int lpos = (bix == bil) ? a.pos(rl) : 0;
+				int len = (bix == biu) ? a.pos(ru - 1) - lpos + a.getCumODims(0) : a.blockSize(bix) * a.getCumODims(0);
+				for (int i = 0; i < len; i++)
+					sum += UtilFunctions.toLong(a.get(bix * a.blockSize() * a.getCumODims(0), lpos + i));
```

Comment:
Did you try to iterate in two loops, rows and other dimensions (from cumODims) and doing a simple get(i,j)? Also, why is there a toLong cast in the case that also applies to FP32?

Reply:
with the simple get(i,j) are you refering to the FP64 case?
My mistake, I should of course NOT cast toLong.

Reply:
I was thinking of something like https://github.com/tugraz-isds/systemds/blob/master/src/main/java/org/tugraz/sysds/runtime/data/TensorBlock.java#L371 (except that we don't even need a switch if we simply compute a double sum) as a fallback for all data types (except string) where we do not want to integrate a custom implementation.